### PR TITLE
ci: Implement basic python linting and autoformatter

### PR DESCRIPTION
**Pull request by alayander**
Implement shell file to run pylint and black autoformatter

### CODE REVIEW

**Comment by alayander** [comment]
In addition to/instead of running black (autoformatter) in the shell script, I can also add git hooks that will run the autoformatter pre-push (on every push attempt) or pre-commit (on every commit attempt)

File: run_ci.sh
Comment (BrianShTsoi) [inline run_ci.sh]: We can add a conditional to check if `ci_venv` is already created. If it is, then we don't have to run all the installation again since presumably it's already set up.

File: run_ci.sh
Comment (BrianShTsoi) [inline run_ci.sh]: Considering we are mostly using macOS and Windows, would it be more straight forward to check if python is installed, and if not we echo instructions telling the user to go to Python's website and install python with the gui installer? Technically macOS can use homebrew, but it is not installed by default. Windows must use the gui installer.
I think telling users to run the installer themselves is also not as intrusive as demanding `sudo` and proceed to implicitly install a bunch of stuff of your machine.

Chatgpt gave me this bash function to detect Python's installation. It works on Git Bash (and WSL)
```
check_python_installed() {
    if command -v python3 &> /dev/null; then
        echo "Python is installed."
    else
        echo "Python is NOT installed."
        echo_python_installation_instructions
    fi
}
```

**Review by BrianShTsoi — CHANGES_REQUESTED** [review]
Some nitpicking to make the script compatible with more OS and also avoid unnecessary commands.

**Comment by alayander** [comment]
@BrianShTsoi Addressed nitpicks, changes requested applied - checks now exist for pre-existing python installation (with exit and prompt for installation otherwise) as well as ci_venv being already set up. Will likely add argument in future to force venv to be recreated since additional packages may be added later as ci suite matures. PTAL

**Comment by alayander** [comment]
> In addition to/instead of running black (autoformatter) in the shell script, I can also add git hooks that will run the autoformatter pre-push (on every push attempt) or pre-commit (on every commit attempt)

What do we think about this idea?

**Comment by BrianShTsoi** [comment]
The changes look good. 

For git hooks yea I think it's a good idea. Pre-commit probably makes sense.

**Comment by BrianShTsoi** [comment]
> Will likely add argument in future to force venv to be recreated since additional packages may be added later as ci suite matures. 

Perhaps we can add a conditional to check if the required packages are available, and if not just install the package?

**Review by BrianShTsoi — CHANGES_REQUESTED** [review]
I'll approve once we have git hooks working as well.

**Comment by alayander** [comment]
Implemented git hook for pre-commit

- I decided to opt for having another script that would fetch and download the pre-commit from a gist: https://gist.github.com/alayander/c7481150e7c2daf0a38dfa05dfd4d982, since I felt this was the least intrusive way to do so.
- This script will need to be re-run to update the pre-commit (as ci suite improves with more autoformatting)

**Comment by alayander** [comment]
> > Will likely add argument in future to force venv to be recreated since additional packages may be added later as ci suite matures.
> 
> Perhaps we can add a conditional to check if the required packages are available, and if not just install the package?

I can address this within the ci_venv.sh file to do checks of what packages are missing per use case (ci script vs pre-commit). This can be part of a separate PR if packages required change as ci suite becomes more mature.

**Review by BrianShTsoi — COMMENTED** [review]
I have two maybe stupid questions:
1. Why do we need `curl` and `gist` in `setup.sh` instead of using a local script?
2. Is it difficult to incorporate the linter into git hooks as well? Adding linter will avoid the mental overhead of the fact that the script for git hooks does something different than `run_ci.sh`

**Comment by alayander** [comment]
> I have two maybe stupid questions:
> 
> 1. Why do we need `curl` and `gist` in `setup.sh` instead of using a local script?
> 
> 2. Is it difficult to incorporate the linter into git hooks as well? Adding linter will avoid the mental overhead of the fact that the script for git hooks does something different than `run_ci.sh`

1. Personally, I thought it was a cleaner decision to have the script maintained separately in a `gist`, but if we want to track that within the repo, I can shift it locally and do some file organization

2. I made this choice since most of the experience I have with git hooks implement autoformatters, but not linters. As well the idea was if black made changes on any files with staged changes, the commit would not be made, but with the linter there's more subjectivity/gray area. Also the `run_ci.sh` script runs over all source code in repo, whereas pre-commit hook only runs on the staged changes wanting to be committed. For this reason it's necessary that the pre-commit hook runs a different script.

**Comment by BrianShTsoi** [comment]
> 1. Personally, I thought it was a cleaner decision to have the script maintained separately in a gist, but if we want to track that within the repo, I can shift it locally and do some file organization

Hmm I guess the benefit of having a gist instead of local file is that users can update the script without the need to pull/rebase? Is that what you mean by cleaner?

> 2. I made this choice since most of the experience I have with git hooks implement autoformatters, but not linters. As well the idea was if black made changes on any files with staged changes, the commit would not be made, but with the linter there's more subjectivity/gray area. Also the run_ci.sh script runs over all source code in repo, whereas pre-commit hook only runs on the staged changes wanting to be committed. For this reason it's necessary that the pre-commit hook runs a different script.

I'm just wondering if we keep them as separate scripts, when do we expect users to use `run_ci.sh`? If we expect them to manually run it pre-commit anyways then it would sort of defeat the purpose of git hooks?
For the subjectivity issue, maybe we can set a standard (something like it has to be at least 9.5/10) and only commit them if it passes the standard? And also add the option to bypass the linter if the user disagrees with the linter's judgement?

**Comment by alayander** [comment]
> Hmm I guess the benefit of having a gist instead of local file is that users can update the script without the need to pull/rebase? Is that what you mean by cleaner?

Yeah that and also I didn't want to clutter the repo with a lot of scripts, however I do see the benefit of tracking the pre-commit within the repo locally, so I can pull that into here (off the gist) and reorganize the added files a bit. 

> I'm just wondering if we keep them as separate scripts, when do we expect users to use `run_ci.sh`? If we expect them to manually run it pre-commit anyways then it would sort of defeat the purpose of git hooks? For the subjectivity issue, maybe we can set a standard (something like it has to be at least 9.5/10) and only commit them if it passes the standard? And also add the option to bypass the linter if the user disagrees with the linter's judgement?

I think it serves as an additional tool; It allows the user to run the ci suite during development before staging any changes to get a first pass of compliance on their code. As for running the linter as well in the pre-commit hook, I see the benefit of having it run and setting a standard with a bypass argument, so I can implement that too.

**Comment by BrianShTsoi** [comment]
The reason I brought those up is that I feel like it'll be ideal to keep it as minimalist/straightforward as possible, both for the user and the maintainer, but I'll leave it up to you.

**Comment by alayander** [comment]
Requested changes are all implemented, although we can't really attach an argument directly to a pre-commit so we can bypass a score below 9 by adding `--no-verify` to git commit command. Workflow wise I think this is fine, since black changes will cause the pre-commit to exit before pylint, so the auto-formatting will need to be staged anyways. And then after reattempting a commit, the pylint output will be displayed and judgement can be used to bypass using the flag if necessary.

**Review by BrianShTsoi — APPROVED** [review]
LGTM!

**Pull request merged**
